drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/PaymentMethodsListDelegate.kt

```diff
@@ -0,0 +1,198 @@
+                order?.paymentMethods?.mapToGiftCardPaymentMethodModel().orEmpty()
+            if (giftCardsList.isNotEmpty()) {
+                add(PaymentMethodHeader(PaymentMethodHeader.TYPE_GIFT_CARD_HEADER))
+                addAll(giftCardsList)
+            }
+            // payment notes
+            order?.remainingAmount?.let { remainingAmount ->
+                val value = CurrencyUtils.formatAmount(remainingAmount, dropInConfiguration.shopperLocale)
+                add(
+                    PaymentMethodNote(application.getString(R.string.checkout_giftcard_pay_remaining_amount, value))
+                )
+            }
+            // stored payment methods
+            storedPaymentMethodsList = storedPaymentMethods.setupStoredPaymentMethods().toMutableList()
+            val hasStoredPaymentMethods = storedPaymentMethodsList.isNotEmpty()
+            if (hasStoredPaymentMethods) {
+                add(PaymentMethodHeader(PaymentMethodHeader.TYPE_STORED_HEADER))
+                addAll(storedPaymentMethodsList)
+            }
+            // payment methods
+            val paymentMethodsList: List<PaymentMethodModel> = paymentMethods.setupPaymentMethodsList()
+            if (paymentMethodsList.isNotEmpty()) {
+                val headerType = if (hasStoredPaymentMethods)
+                    PaymentMethodHeader.TYPE_REGULAR_HEADER_WITHOUT_STORED
+                else
+                    PaymentMethodHeader.TYPE_REGULAR_HEADER_WITH_STORED
+                add(PaymentMethodHeader(headerType))
+                addAll(paymentMethodsList)
+            }
+        }
+        _paymentMethodsFlow.tryEmit(paymentMethods)
+    }
+
+    internal fun removePaymentMethodWithId(id: String) {
+        if (this::storedPaymentMethodsList.isInitialized) {
+            val item = storedPaymentMethodsList.first { item -> item.id == id }
+            storedPaymentMethodsList.remove(item)
+            populatePaymentMethods()
+        } else throw IllegalStateException("store payment methods not found")
+    }
+
+    private fun List<PaymentMethod>.setupPaymentMethodsList(): List<PaymentMethodModel> =
+        this.mapIndexedNotNull { index, paymentMethod ->
+            val isAvailable = checkPaymentMethodsMap[paymentMethod]
+                ?: throw IllegalStateException("payment method not found in map")
+            if (isAvailable) {
+                paymentMethod.mapToModel(index)
+            } else
+                null
+        }
+
+    private fun List<StoredPaymentMethod>.setupStoredPaymentMethods(): List<StoredPaymentMethodModel> =
+        this.mapNotNull { storedPaymentMethod ->
+            if (storedPaymentMethod.isStoredPaymentSupported())
+                storedPaymentMethod.mapStoredModel(dropInConfiguration.isRemovingStoredPaymentMethodsEnabled)
+            else
+                null
+        }
+
+    private fun StoredPaymentMethod.isStoredPaymentSupported(): Boolean {
```

Comment:
This can be moved to `StoredUtils`, and used in other places as well, check out `DropInViewModel.shouldShowPreselectedStored` and `DropInViewModel.getPreselectedStoredPaymentMethod`

Reply:
nice idea